### PR TITLE
Small blog post on using grouping for result diversification

### PR DESCRIPTION
**Pull request by jobergum**
Try to address questions on how to achieve result diversity with Vespa. 

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by jobergum** [comment]
image

File: _posts/2021-11-11-result-diversification-with-vespa.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+The grouped result computed in parallel over all cluster content nodes can be
+post-processed in a [stateless
+searcher](https://docs.vespa.ai/en/searcher-development.html). 
+Then, the developer can further
+diversify the result using custom business logic and finally present the
+diversified results.  For example, when ordering groups by the max relevancy and
+emitting more than one document per group, it
+makes sense to check the relevance of the secondary hits from the group.
+The searcher can also build and process the grouping request and response, see
+[Searcher grouping api](https://docs.vespa.ai/en/grouping.html#search-container-api)
```
Comment (kkraune) [inline _posts/2021-11-11-result-diversification-with-vespa.md:147]:
```suggestion
[Searcher grouping api](https://docs.vespa.ai/en/grouping.html#search-container-api).
```

File: _posts/2021-11-11-result-diversification-with-vespa.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+# Serving performance  
+Four main components drive serving performance when the query request includes
+result grouping. In order of importance:
+
+- The number of matches the query produces per node. All the document matches
+  get
+exposed to the grouping framework. The total result hit count of the query is
+equal to the number of hits exposed to grouping. 
+- The total number of unique values the field can take.
```
Comment (kkraune) [inline _posts/2021-11-11-result-diversification-with-vespa.md:157]:
```suggestion
- The total number of unique values in the field. 
```

File: _posts/2021-11-11-result-diversification-with-vespa.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+It's also possible to limit the number of documents exposed to grouping by using 
+the [match-phase
+degradation](https://docs.vespa.ai/en/graceful-degradation.html#match-phase-degradation)
+feature, which sets an upper limit on the number of documents ranked
+, using a document attribute. This feature also includes diversity criteria
+so that the results exposed to grouping also are diversified.
+
+The number of unique values the field can take is data-dependent. A few ten
+thousand unique values is usually a breeze. The number of nodes in the cluster
+to which the query is fan out increases network bandwidth. The performance
```
Comment (kkraune) [inline _posts/2021-11-11-result-diversification-with-vespa.md:191]:
```suggestion
to which the query is fanned out increases network bandwidth. The performance
```

File: _posts/2021-11-11-result-diversification-with-vespa.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+It is possible to order groups by more complex expressions working on
+aggregates like <em>sum()</em> and <em>count()</em>, for example:
+- Number of hits in the group times the maximum relevance:
+<em>order(-max(relevance())*count())</em> 
+- The sum of a document ctr attribute for the hits in the group times the
+  maximum
+relevance: <em>order(-max(relevance())*sum(ctr))</em>
+
+# Fine-tuning result diversification
```
Comment (kkraune) [inline _posts/2021-11-11-result-diversification-with-vespa.md:137]: As the blog post is about diversification, we should add a few more details here, like a more real-world example with some pseudo-code

**Review by kkraune — CHANGES_REQUESTED** [review]
Great post, a few comments!

File: _posts/2021-11-11-result-diversification-with-vespa.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+</pre>
+
+In this expression, <em>K</em> is the maximum number of matches per
+node that grouping runs over. 
+
+It's also possible to limit the number of <em>matches</em> exposed to grouping by using 
+the [match-phase
+degradation](https://docs.vespa.ai/en/graceful-degradation.html#match-phase-degradation)
+feature, which sets an upper limit on the number of documents ranked
+, using a document side quality attribute. The match phase degradation feature supports diversity criteria
```
Comment (kkraune) [inline _posts/2021-11-11-result-diversification-with-vespa.md:251]:
```suggestion
feature, which sets an upper limit on the number of documents ranked,
using a document side quality attribute. The match phase degradation feature supports diversity criteria
```

**Review by kkraune — APPROVED** [review]
This is now a great post - ready to go with the minor fixes to punctuation!

**Pull request merged**